File: src/domain/kTokens.ts
import { Effect } from "effect"
import type { TaskError } from "./types.js"

export interface KTokensConfig {
  anchorPoint: number      // Fibonacci number to anchor against
  kTokensAtAnchor: number  // how many kTokens map to anchorPoint
  maxKTokens: number       // cap (inclusive)
}

export const defaultConfig: KTokensConfig = {
  anchorPoint: 8,
  kTokensAtAnchor: 20,
  maxKTokens: 20,
}

// Fibonacci sequence for lookup
const FIBONACCI = [1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144]

export const estimateFromKTokens = (
  kTokens: number,
  config: KTokensConfig = defaultConfig,
): Effect.Effect<number, TaskError> => {
  // Fail if kTokens <= 0
  if (kTokens <= 0) {
    return Effect.fail({
      _tag: 'validation_error',
      message: 'predicted kilotokens must be positive',
    })
  }

  // Fail if kTokens > config.maxKTokens
  if (kTokens > config.maxKTokens) {
    return Effect.fail({
      _tag: 'validation_error',
      message: 'predicted kilotokens exceed maximum allowed',
    })
  }

  // Calculate ratio: kTokensAtAnchor / anchorPoint
  const ratio = config.kTokensAtAnchor / config.anchorPoint

  // Scale kTokens to estimate space
  const scaled = kTokens / ratio

  // Find nearest Fibonacci number, rounding UP on tie
  let nearestFib: number = FIBONACCI[0] ?? 1
  let minDistance = Math.abs(nearestFib - scaled)

  for (const fib of FIBONACCI) {
    const distance = Math.abs(fib - scaled)

    // On exact match, return immediately
    if (distance === 0) {
      return Effect.succeed(fib)
    }

    // On tie, pick the larger value (UP)
    if (distance < minDistance || (distance === minDistance && fib > nearestFib)) {
      nearestFib = fib
      minDistance = distance
    }
  }

  return Effect.succeed(nearestFib)
}

File: internal/logbook/ktokens.go
package logbook

func EstimateFromKTokens(kTokens int) (int, error) {
	if kTokens <= 0 {
		return 0, TaskError{Tag: "validation_error", Message: "predicted kilotokens must be positive"}
	}
	if kTokens > 20 {
		return 0, TaskError{Tag: "validation_error", Message: "predicted kilotokens exceed maximum allowed"}
	}
	ratio := 20.0 / 8.0
	scaled := float64(kTokens) / ratio
	return nearestFib(scaled), nil
}

func ValidateFibonacci(n int) error {
	if n <= 0 {
		return TaskError{Tag: "validation_error", Message: "estimation must be a Fibonacci number"}
	}
	five := 5 * n * n
	if isPerfectSquare(five+4) || isPerfectSquare(five-4) {
		return nil
	}
	return TaskError{Tag: "validation_error", Message: "estimation must be a Fibonacci number"}
}

func isPerfectSquare(n int) bool {
	if n < 0 {
		return false
	}
	r := int64Sqrt(int64(n))
	return r*r == int64(n)
}

func int64Sqrt(n int64) int64 {
	if n <= 0 {
		return 0
	}
	x := n
	y := (x + 1) / 2
	for y < x {
		x = y
		y = (x + n/x) / 2
	}
	return x
}
